File: 0322-coinChange.java
// https://leetcode.com/problems/coin-change/

class Solution {
    private static final int INVALID = -1;
    
    public int coinChange(int[] coins, int amount) {
        int[] memo = new int[amount + 1];
        Arrays.fill(memo, -2);
        
        return step(coins, amount, memo);
    }
    
    private int step(int[] coins, int balance, int[] memo) {
        if (balance < 0) {
            return INVALID;
        } else if (balance == 0) {
            return 0;
        } else if (memo[balance] != -2) {
            return memo[balance];
        }
        
        int best = INVALID;
        
        for (int coin : coins) {
            int next = step(coins, balance - coin, memo);
            if (next == INVALID) {
                continue;
            }
            
            int curr = 1 + next;
            if (best == INVALID || curr < best) {
                best = curr;
            }
        }
        
        memo[balance] = best;
        return best;
    }
}

File: 0322-coinChange.js
// https://leetcode.com/problems/coin-change/

function coinChange(coins, amount) {
    coins.sort((a, b) => b - a);
    
    const memo = new Array(amount + 1).fill(-1);
    
    function step(bal) {
        if (bal === 0) {
            return 0;
        } else if (bal < 0) {
            return Infinity;
        } else if (memo[bal] !== -1) {
            return memo[bal];
        }
        
        let best = Infinity;
        
        for (const coin of coins) {
            const curr = 1 + step(bal - coin);
            best = Math.min(best, curr);
        }
        
        memo[bal] = best;
        return best;
    }
    
    const result = step(amount);
    return (result !== Infinity)
        ? result
        : -1;
}